In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
# Load dataset Iris dari scikit-learn
iris = load_iris()


Cell ini mengimport NumPy (untuk komputasi numerik dan array), Pandas (untuk mengolah data dalam bentuk DataFrame), dan fungsi load_iris dari Scikit-Learn. Baris iris = load_iris() memuat dataset Iris yang berisi 150 data bunga dengan 4 fitur numerik dan 3 kelas spesies.


In [2]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]
# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["Species"])
y = df["Species"]
print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


Data Iris diubah menjadi DataFrame dengan nama kolom sesuai feature_names. Kolom "Species" ditambahkan dengan mengubah label angka (iris.target) menjadi nama spesies lewat iris.target_names. Setelah itu data dipisah menjadi X (4 fitur: sepal length, sepal width, petal length, petal width) dan y (label spesies). Output print(X) menunjukkan 150 baris × 4 kolom.


In [3]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

Menampilkan isi y, yaitu label spesies setiap data. Data tersusun berurutan per kelas: indeks 0–49 setosa, 50–99 versicolor, dan 100–149 virginica. Totalnya 150 label bertipe teks (object), sehingga perlu diubah menjadi angka.

In [4]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

LabelEncoder mengubah label teks menjadi angka berdasarkan urutan abjad: setosa = 0, versicolor = 1, virginica = 2. Hasilnya array berisi 50 angka untuk masing-masing kelas. Objek lab disimpan agar nanti hasil prediksi berupa angka bisa dikembalikan ke nama spesies dengan inverse_transform.

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)


Data dibagi dengan test_size=0.2, yaitu 80% (120 data) untuk melatih model dan 20% (30 data) untuk menguji model. random_state=0 membuat pengacakan selalu sama setiap kali dijalankan. Karena X dan y dibagi dengan random_state yang sama, pasangan fitur dan labelnya tetap sesuai.

In [6]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
NB_model = model.fit(X_train, y_train)


GaussianNB diimpor lalu dilatih dengan fit(X_train, y_train). Berbeda dengan KNN yang hanya menyimpan data, di tahap ini Naïve Bayes benar-benar "belajar". Model menghitung prior P(C) setiap kelas, serta rata-rata (μ) dan standar deviasi (σ) setiap fitur di setiap kelas. Nilai-nilai inilah yang nanti dipakai di rumus Gaussian untuk menghitung P(X|C).

In [7]:
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

Model memprediksi 30 data uji. Untuk setiap data, model menghitung probabilitas posterior tiap kelas dengan mengalikan prior P(C) dengan P(X|C) dari keempat fitur (asumsi fitur saling independen), lalu memilih kelas dengan probabilitas tertinggi. Hasilnya array 30 prediksi dalam bentuk angka 0/1/2.

In [8]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)


Accuracy of data testing:  96.66666666666667


accuracy_score membandingkan hasil prediksi dengan label sebenarnya, lalu dikali 100 agar tampil dalam persen. Akurasinya 96,67%, artinya 29 dari 30 data uji benar dan hanya 1 yang salah. Nilai ini sama dengan akurasi KNN (K = 3) pada praktikum sebelumnya, sehingga kedua metode sama-sama bekerja sangat baik pada dataset Iris.

In [9]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


d:\Python311\Lib\site-packages\sklearn\utils\validation.py:2830: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


Model memprediksi data baru [5.1, 3.5, 1.4, 0.2], lalu inverse_transform mengubah hasil angka menjadi nama spesies. Hasilnya setosa, karena petal length dan petal width-nya sangat kecil, ciri khas setosa. Muncul UserWarning karena model dilatih dengan DataFrame yang punya nama kolom, sedangkan input diberikan sebagai list biasa tanpa nama kolom. Warning ini tidak memengaruhi hasil prediksi, hanya peringatan bahwa struktur datanya tidak konsisten.

In [10]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


Input yang sama dibungkus dengan pd.DataFrame(..., columns=iris.feature_names) sehingga memiliki nama kolom yang sama dengan data latih. Warning tidak muncul lagi, dan hasil prediksinya tetap setosa. Cara inilah yang sebaiknya dipakai setiap kali memprediksi data baru.

### Tugas

In [11]:
# Tugas 2: prediksi data baru
input_data = pd.DataFrame(np.array([[7.3, 3.8, 4.6, 2.3]]), columns=iris.feature_names)

result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

# untuk melihat probabilitas tiap kelas
print('Probabilitas [setosa, versicolor, virginica]:', NB_model.predict_proba(input_data).round(4))

Hasil prediksi untuk input baru adalah: ['virginica']
Probabilitas [setosa, versicolor, virginica]: [[0. 0. 1.]]


Data baru dibuat sebagai DataFrame dengan nama kolom yang sesuai, diprediksi dengan model Gaussian NB, lalu dikembalikan ke nama spesies dengan inverse_transform. Hasilnya virginica. predict_proba menampilkan probabilitas tiap kelas [[0. 0. 1.]], artinya model hampir 100% yakin datanya virginica. Angka 0 pada setosa dan versicolor sebenarnya bukan nol persis, tapi sangat kecil sehingga terbulatkan.

Alasannya, petal width 2,3 sangat dekat dengan rata-rata virginica (2,03) dan jauh dari versicolor (1,32) maupun setosa (0,24), dan sepal length 7,3 juga paling sesuai dengan virginica. Petal length 4,6 sebenarnya lebih dekat ke versicolor (rata-rata 4,22), tapi karena Naïve Bayes mengalikan probabilitas semua fitur, satu fitur yang sangat tidak cocok (petal width) sudah cukup membuat probabilitas versicolor jatuh sangat kecil.
